flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,533 @@
+  }
+
+  /**
+    * Register a timer for cleaning up records in a specified time.
+    *
+    * @param ctx               the context to register timer
+    * @param timeForRecord     time for the input record
+    * @param oppositeWatermark watermark of the opposite stream
+    * @param timerState        stores the timestamp for the next timer
+    * @param leftRecord        record from the left or the right stream
+    * @param firstTimer        whether this is the first timer
+    */
+  private def registerCleanUpTimer(
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    timeForRecord: Long,
+    oppositeWatermark: Long,
+    timerState: ValueState[Long],
+    leftRecord: Boolean,
+    firstTimer: Boolean): Unit = {
+    val cleanUpTime = timeForRecord + (if (leftRecord) leftRelativeSize else rightRelativeSize) +
+      allowedLateness + 1
+    registerTimer(ctx, !leftRecord, cleanUpTime)
+    LOG.debug(s"Register a clean up timer on the ${if (leftRecord) "RIGHT" else "LEFT"} state:"
+      + s" timeForRecord = ${timeForRecord}, cleanUpTime = ${cleanUpTime}, oppositeWatermark = " +
+      s"${oppositeWatermark}")
+    timerState.update(cleanUpTime)
+    if (cleanUpTime <= oppositeWatermark + allowedLateness && firstTimer) {
+      backPressureSuggestion =
+        if (leftRecord) (oppositeWatermark + allowedLateness - cleanUpTime)
+        else -(oppositeWatermark + allowedLateness - cleanUpTime)
+      LOG.warn("The clean timer for the " +
+        s"${if (leftRecord) "left" else "right"}" +
+        s" stream is lower than ${if (leftRecord) "right" else "left"} watermark." +
+        s" requiredTime = ${formatTime(cleanUpTime)}, watermark = ${formatTime(oppositeWatermark)},"
+        + s"backPressureSuggestion = " + s"${backPressureSuggestion}.")
+    }
+  }
+
+
+  /**
+    * Called when a registered timer is fired.
+    * Remove records which are earlier than the expiration time,
+    * and register a new timer for the earliest remaining records.
+    *
+    * @param timestamp the timestamp of the timer
+    * @param ctx       the context to register timer or get current time
+    * @param out       the collector for returning result values
+    */
+  override def onTimer(
+    timestamp: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    out: Collector[CRow]): Unit = {
+    getCurrentOperatorTime(ctx)
+    //TODO In the future, we should separate the left and right watermarks. Otherwise, the
+    //TODO registered timer of the faster stream will be delayed, even if the watermarks have
+    //TODO already been emitted by the source.
+    if (leftTimerState.value == timestamp) {
+      val rightExpirationTime = leftOperatorTime - rightRelativeSize - allowedLateness - 1
+      removeExpiredRecords(
+        timestamp,
+        rightExpirationTime,
+        leftOperatorTime,
+        rightCache,
+        leftTimerState,
+        ctx,
+        false
+      )
+    }
+
+    if (rightTimerState.value == timestamp) {
```

Comment:
can be the `else` case without condition, right?

Reply:
I am thinking if the `rightState.value` could be equal with the `leftState.value`.

Reply:
Yes, you are right.